File: Data Structures & Algorithms/house-robber/submission-4.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # to rob good we need to rob the maximum betwen the last 2 before me 
        if len(nums) == 0 :
            return 0
        if len(nums) == 1 :
            return nums[0]            
        for index , num in enumerate(nums):
            if index == 0 or index == 1 : 
                continue 
            if index == 2 :
                nums[index] = max(nums[index - 2] + num , nums[index - 1] )
                continue                
            nums[index] = max(nums[index - 2] + num , nums[index - 1] , nums[index - 3 ] + num)
        print(nums)            
        return  max (nums[-1], nums[1] , nums[0])                             

        

File: Data Structures & Algorithms/top-k-elements-in-list/submission-4.java

class Solution {
    public int[] topKFrequent(int[] nums, int k) {

        HashMap<Integer, Integer> map = new HashMap<>();
        
        for (int i = 0; i < nums.length; i++) {
            if (map.get(nums[i]) == null) {
                map.put(nums[i], 1);
            }
            else {
                map.put(nums[i], map.get(nums[i]) + 1);
            }
        }
        
        LinkedList<Integer>[] freq = new LinkedList[nums.length];

        for(int i = 0; i < nums.length; i++) {
            freq[i] = new LinkedList<>();
        }
        
        map.forEach((key, value) -> {
            freq[value - 1].add(key);
        });

        int[] max = new int[k];
        int index = 0;
        for (int i = freq.length - 1; i >= 0 ; i--) {
            if (freq[i].size() == 0)
                continue;
            
            for(Integer val : freq[i]) {
                max[index] = val;
                if (index == k - 1)
                    return max;
                index++;
            }
        }
        return max;
    }
}
